### PR TITLE
Low-level interop for PG arrays

### PR DESCRIPTION
**Pull request by mhov**
These PG function shims will allow for future work to improve the `pgx::Array<T>` type, as well as manual access to ArrayType data in the meantime.

### CODE REVIEW

File: pgx-pg-sys/cshim/pgx-cshim.c
Comment (workingjubilee) [inline pgx-pg-sys/cshim/pgx-cshim.c:133]: Hmm. Shouldn't we also directly expose `ARR_DIMS`, as it might be desired by Rust code?

I thought `ARR_NDIM(ArrayType*)` was just a direct field read,
so if I remember my C, `ARR_NELEMS` could be written:
```c
int pgx_ARR_NELEMS(ArrayType *arr) {
    return ArrayGetNItems(arr->ndim, ARR_DIMS(arr));
}
```

File: pgx-pg-sys/src/lib.rs
Comment (workingjubilee) [inline pgx-pg-sys/src/lib.rs:457]: This is not a safe function. It is unsound to call this if `*mut ArrayType` is the null pointer or an invalid pointer, as it involves **reading a field** of the pointee to determine what to do next. A function must **not** be safe if it is memory-unsafe to call it with **any** valid input for that type, **especially** including the null pointer.

This function may be desirable but it **must** be `unsafe`.

File: pgx-pg-sys/src/lib.rs
Comment (workingjubilee) [inline pgx-pg-sys/src/lib.rs:488]: I believe none of these are actually safe due to the prior reason: a field read on a possibly-invalid, possibly-null pointer happens.

**Review by workingjubilee — CHANGES_REQUESTED** [review]
Do the helper functions have to live in `pgx-pg-sys`? Because they should be purely additive to the sys crate, I would somewhat prefer they were exposed through a `mod` in the main `pgx` crate, so as to make it easier to thread them through the other code in `pgx` without violating coherence.

I also would like it if all new `unsafe` code that lands in PGX to document its public `/// # Safety` requirements as precisely as possible, and for every invocation of `unsafe` to come with `// SAFETY:` annotations,  and I would prefer it to come with `#[deny(unsafe_op_in_unsafe_fn)]` on the containing module so as to require **both of those** even inside an `unsafe fn`.

It is the last one, the extreme safety requirements, that is completely impractical to implement throughout `pgx-pg-sys`. Note that the comments I desire can be **very** terse, they do not have to be elaborate, they just have to make clear all the little details like "this does a field read".

**Comment by workingjubilee** [comment]
The burden of assuring memory safety must end the moment code leaves an `unsafe` context. This means that a safe function may not impose any requirements whatsoever on its callers beyond that which it imposes on the purely programmatic level. It is permissible to induce logic errors for illogical inputs, but trying to read values out of the null page (or, really, any field offset from the null pointer) is definitely not one of them.

e.g. It is fine if an impl of `Ord` is incorrect for some comparisons, because Ord is safe code. Thus when your `sort` fn relies on an impl of `Ord` to be correct in order to return a correct sort, it is fine if the resulting sort is wrong.

It is not fine if you then rely on the impl of `Ord` being correct for memory safety purposes. This is an actual, in-the-wild concern in real code that is not Rust, as glibc's `qsort` in the past has induced segfaults on incorrect comparison functions. Rust does not tolerate that kind of shenanigans, so glibc's `qsort` would be considered `unsafe`. The direct implication is that safe code is allowed to "shit the bed", from the perspective of `unsafe` code, and the `unsafe` code must either pass the burden on to a caller, or clean it up. Thus, for a function which provides a safe wrapper for `unsafe` code, **the buck stops there**. It must guarantee all safely-obtained inputs in Rust (and it is "safe" code to write `std::ptr::null_mut()`) to the `unsafe` block inside do not violate memory safety.

...Directly, mind. Unfortunately running `DROP DATABASE IF EXISTS prod` is in the category of things that are considered "logic errors", not "memory safety violations", despite having a marvelous potential to erase data you may have wanted to keep around.

File: pgx/src/array.rs
Comment (workingjubilee) [inline pgx/src/array.rs:103]: This isn't quite right. It's fine to use this if you also "consumed" RawArray to produce the slice.

File: pgx-pg-sys/src/lib.rs
Comment (workingjubilee) [reply pgx-pg-sys/src/lib.rs]: I have introduced safe wrappers for everything except the null bitmap functions, which will require another wave of commits.

**Review by workingjubilee — DISMISSED** [review]
Needs a second pass, as already mentioned.

File: pgx/src/array.rs
Comment (workingjubilee) [reply pgx/src/array.rs:19]: This was verrry incorrect to use, as it went from a possibly null, possibly invalid raw pointer, to a slice.

File: pgx-tests/src/tests/array_tests.rs
Comment (BradyBonnette) [inline pgx-tests/src/tests/array_tests.rs:116]: What is the reason for marking the fn as `unsafe` but also having the body of the function also wrapped in an `unsafe` block? The `get_arr_nelems` function above does not have it. Is it because this fn returns an `Option<>`?

File: pgx-tests/src/tests/array_tests.rs
Comment (BradyBonnette) [inline pgx-tests/src/tests/array_tests.rs:129]: `&*slice.as_ptr()`  👀

File: pgx-tests/src/tests/array_tests.rs
Comment (BradyBonnette) [inline pgx-tests/src/tests/array_tests.rs:328]: Is this to show which indexes have nulls and which indexes have non-null (apologies for the crude ascii diagram that may not render correctly)?

```
0b00010101
    [5n3n1]
```

File: pgx/src/array.rs
Comment (BradyBonnette) [inline pgx/src/array.rs:23]: 👀 Had to look up `repr(transparent)`

File: pgx/src/array.rs
Comment (BradyBonnette) [inline pgx/src/array.rs:133]: `// However, PGX has trouble with returning usizes`

I saw this comment earlier. Is this something that'll need to be addressed at some point? Can it even be addressed?

File: pgx/src/array.rs
Comment (BradyBonnette) [inline pgx/src/array.rs:152]: `MIGHT` 🤔

File: pgx/src/array.rs
Comment (BradyBonnette) [inline pgx/src/array.rs:192]: What's `UB` here? "Upper bound"?

File: pgx/src/array.rs
Comment (BradyBonnette) [inline pgx/src/array.rs:195]: Does using `usize` here have any implications, since earlier it was mentioned that PGX has issues returning `usize`?

File: pgx-tests/src/tests/array_tests.rs
Comment (workingjubilee) [reply pgx-tests/src/tests/array_tests.rs:116]: It's because I think we should assume that if a Postgres varlena array has "nullable indices", that we should treat the resulting slice as [`&[MaybeUninit<T>]`](https://doc.rust-lang.org/stable/std/mem/union.MaybeUninit.html) so it is incorrect to read the value of an index that was nulled.

The latter is because I want to move towards `#[deny(unsafe_op_in_unsafe_fn)]` across the repository.

File: pgx/src/array.rs
Comment (workingjubilee) [reply pgx/src/array.rs:133]: Yes, it should be addressed.
I have no idea how best to do so currently.

File: pgx/src/array.rs
Comment (workingjubilee) [reply pgx/src/array.rs:192]: undefined behavior

File: pgx/src/array.rs
Comment (workingjubilee) [reply pgx/src/array.rs:195]: Basically Rust indices by `usize`, but Postgres indices with `c_int`, so we can usually safely upgrade a Postgres index into a Rust index because either the machine is 32-bit or 64-bit and we're not really running on 16-bit machines.

File: pgx-tests/src/tests/array_tests.rs
Comment (workingjubilee) [reply pgx-tests/src/tests/array_tests.rs:328]: Correct. @mhov wrote that test, but it basically prints the bitwise repr of the null bitmap that Postgres uses internally.

File: pgx-tests/src/tests/array_tests.rs
Comment (workingjubilee) [reply pgx-tests/src/tests/array_tests.rs:129]: This pattern is called a "reborrow".
It brings us from `NonNull<[T]>` to `&[T]`.

File: pgx-tests/src/tests/array_tests.rs
Comment (workingjubilee) [reply pgx-tests/src/tests/array_tests.rs:328]: I believe it's actually the reverse though:
```
00001010
   5n3n1
```

File: pgx/src/array.rs
Comment (workingjubilee) [reply pgx/src/array.rs:152]: PERHAPS UNNECESSARY CAPITALIZATION, or?

File: pgx/src/array.rs
Comment (workingjubilee) [reply pgx/src/array.rs:23]: Hm. I'm not sure if it actually should be, tbh. Might want to look up the `len` and cache it on construction.

File: pgx/src/array.rs
Comment (BradyBonnette) [reply pgx/src/array.rs:133]: Ah ok

File: pgx/src/array.rs
Comment (BradyBonnette) [reply pgx/src/array.rs:152]: No, not at all, it's a good callout. It was more like "Im calling something that might have null" instead of "I am calling something I know for sure will be null or not".

File: pgx/src/array.rs
Comment (BradyBonnette) [reply pgx/src/array.rs:192]: I don't know why I keep forgetting that, lol

File: pgx/src/array.rs
Comment (BradyBonnette) [reply pgx/src/array.rs:195]: Ohhhhhhhh ok, I see. Thanks!

File: pgx-tests/src/tests/array_tests.rs
Comment (BradyBonnette) [reply pgx-tests/src/tests/array_tests.rs:129]: Ah, more reading material. Thanks 😄

File: pgx/src/array.rs
Comment (BradyBonnette) [inline pgx/src/array.rs:33]: Thanks for the expanded explanation 😃

File: pgx/src/array.rs
Comment (workingjubilee) [reply pgx/src/array.rs:34]: > Hm. I'm not sure if it actually should be, tbh. Might want to look up the `len` and cache it on construction.

Ah, I believe I will learn what we should do when I construct the followups to this that wrap RawArray further to allow more "fully safe" handling of it. It may require changing some design specs given here.

File: pgx/src/array.rs
Comment (workingjubilee) [inline pgx/src/array.rs:225]: You don't specifically document that it's asserting `[T]`, you're only asking for `T`, and there's some nuances between those.

File: pgx/src/array.rs
Comment (workingjubilee) [inline pgx/src/array.rs:175]: Same question: maybe this should path should be public?

File: pgx/src/array.rs
Comment (workingjubilee) [inline pgx/src/array.rs:112]: Should this path be visible in the public documentation? I suppose they can click through to see the source, but the public docs will support links.

File: pgx/src/array.rs
Comment (workingjubilee) [inline pgx/src/array.rs:49]: Talking about the intent of this type might be useful to publicly discuss in its documentation.

File: pgx/src/array.rs
Comment (workingjubilee) [inline pgx/src/array.rs:23]: You may want a more terse initial comment, I think, for the "snippet form" in the docs.

File: pgx/src/array.rs
Comment (workingjubilee) [inline pgx/src/array.rs:65]: Style elsewhere in the repo has been to use `/*` (block), `/*!` (inner block doc), and `/**` (block doc) for multiline comments, consider applying that here?

**Review by workingjubilee — CHANGES_REQUESTED** [review]
This needs improvement in the documentation before it lands, even though there is going to be a followup PR soon.

File: pgx-tests/src/tests/array_tests.rs
Comment (workingjubilee) [reply pgx-tests/src/tests/array_tests.rs:116]: You should open an issue if you want to do this across the entire repo, @workingjubilee, so other people working with you can help with it.a

Comment (workingjubilee) [reply pgx-tests/src/tests/array_tests.rs:116]: Opened:
- https://github.com/tcdi/pgx/issues/653

File: pgx-tests/src/tests/array_tests.rs
Comment (workingjubilee) [reply pgx-tests/src/tests/array_tests.rs:329]: This tendency in the formatting detail needs to be documented.

Comment (workingjubilee) [reply pgx-tests/src/tests/array_tests.rs:329]: What were you saying, @workingjubilee? It is what Brady just said, as witnessed here in this very test. The nullbitmap uses a 0 to mark "is null", unlike all the rest of Postgres and `NullableDatum`, which uses `true` bools to mark "is null", because Postgres has very good ideas.

File: pgx/src/array.rs
Comment (workingjubilee) [reply pgx/src/array.rs:74]: I was going to change it to not do that NULL-check in a followup PR. Should I fix that now?

File: pgx/src/array.rs
Comment (workingjubilee) [reply pgx/src/array.rs:109]: Mmh. Isn't there a case people can make an incorrect `&[T]` slice if we don't offer this ourselves?

File: pgx/src/array.rs
Comment (workingjubilee) [reply pgx/src/array.rs:112]: Yeah, actually, directly linking is a good idea. And that way we could also link to the doxygen.

File: pgx/src/array.rs
Comment (workingjubilee) [reply pgx/src/array.rs:135]: Well, it's the number of elements in the entire array, covering both those that are invalid (i.e. marked null in the nullbitmap) and valid, without regard to any given dimension.

I guess it's more traditional to call that "size" or something? Postgres uses `nitems`, and elsewhere PGX uses `nelems`.

File: pgx/src/array.rs
Comment (workingjubilee) [reply pgx/src/array.rs:176]: It's just going to be optimized to this anyways, and sometimes `%` and `/` are missed opts by LLVM.

File: pgx/src/array.rs
Comment (workingjubilee) [reply pgx/src/array.rs:74]: Oh nah that's fine, just leave this unresolved until the followup PR, it's nonblocking.

File: pgx/src/array.rs
Comment (workingjubilee) [reply pgx/src/array.rs:139]: Okay, I get what you are saying, I think. We should really not deref the slices we are exposing to users, and only do the absolute minimum possible here. That also weakens the requirements on this type. Then the next type can build on top of this with a safer abstraction, so the nullbitmap ptr and the data ptr both should point to `Option<NonNull<u8>>` and `Option<NonNull<T>` respectively. The user of this type can cache the len and wrap this properly and return slices. I'm just worried about people misusing these pointers.

File: pgx/src/array.rs
Comment (workingjubilee) [reply pgx/src/array.rs:172]: I like the pithy one-word names, tho', and `nullbitmap()` is not really idiomatic for Rust, it would be `null_bitmap`, I think, which feels much worse, and it would be the `nulls_bitmap` anyways, since it describes the indices of `nulls`, and then you see how this got back to just `nulls`, yes?

File: pgx/src/array.rs
Comment (workingjubilee) [reply pgx/src/array.rs:109]: Yeah, this should remain as-is, and actually `ndims` is probably the fn that should go, because they can call `dims().len()`

File: pgx/src/array.rs
Comment (workingjubilee) [reply pgx/src/array.rs:148]: Ugh, the results of this are too painful to use. Having two types for the same low-level accessors feels a bit silly???

File: pgx/src/array.rs
Comment (workingjubilee) [inline pgx/src/array.rs:248]: This was a cute idea for a few seconds but then you dropped the PhantomData type. These are definitely not different things.

File: pgx/src/array.rs
Comment (workingjubilee) [inline pgx/src/array.rs:290]: I don't think this needs to be unsafe if it loads the requirements into construction.

**Review by workingjubilee — CHANGES_REQUESTED** [review]

File: pgx/src/array.rs
Comment (workingjubilee) [reply pgx/src/array.rs]: Really? Sad.

File: pgx/src/array.rs
Comment (workingjubilee) [reply pgx/src/array.rs:188]: Fair enough.

File: pgx/src/array.rs
Comment (workingjubilee) [reply pgx/src/array.rs]: Well, also, invalid pointers aren't unsafe to construct, yadda yadda.

File: pgx/src/array.rs
Comment (workingjubilee) [reply pgx/src/array.rs]: Doxygen is too unstable, will point to master and lines might change. Instead used gitweb links.

File: pgx/src/array.rs
Comment (BradyBonnette) [reply pgx/src/array.rs:211]: In my mind, if the intention is for it to be used as a bitmap by the caller (as opposed to a "regular" map), then I don't think there's anything inherently wrong with having the world "bitmap" in there somewhere. It at least makes the intention clear, and a few extra characters probably won't hurt anything.

I don't think I can necessarily speak to how idiomatic either/all names would be, but having a clear intention is nice.

Though on second look, the one thing that did sorta confuse me (as an early-stage Rust developer, mind) is that the intention is "this is a bitmap of things which indicate that something is Null" but the Option type is `NonNull`. I get that `NonNull` is a pointer type (I think?), and more seasoned Rust developers would understand that in a heartbeat. It just sorta threw me off for a quick second.

File: pgx/src/array.rs
Comment (BradyBonnette) [reply pgx/src/array.rs]: How much detail would be required to make it "make sense" to a newcomer without also making it overbearing to read? Would a consumer of this need to fully understand what Postgres does with it?

File: pgx/src/array.rs
Comment (BradyBonnette) [reply pgx/src/array.rs]: I don't think it'd hurt, and having it reference Postgres' source is nice to know. Not as in an actual hyperlink that could change over time, but just a general "look here" as you have I think is sufficient. Actually there was a point in time while reviewing this stuff I looked at the Postgres source mirror on Github just to reference what you had written.

File: pgx/src/array.rs
Comment (BradyBonnette) [reply pgx/src/array.rs]: I kinda wonder if your answer to your own question should also be in the comments?  😃

File: pgx/src/array.rs
Comment (BradyBonnette) [reply pgx/src/array.rs]: From a pgx-maintainers perspective, how much of a maintenance nightmare is it linking to the doxygen pages? How often would that change?

Comment (workingjubilee) [reply pgx/src/array.rs]: Potentially every time someone lands a commit into postgresql/master.

File: pgx/src/array.rs
Comment (workingjubilee) [inline pgx/src/array.rs:116]: Stray big multiline comment with `//`

**Review by workingjubilee — APPROVED** [review]
Wow, that was... quite a few commits, and only 11% were rustfmt commits! Nicely done!

File: pgx/src/array.rs
Comment (workingjubilee) [inline pgx/src/array.rs:165]: You ain't gonna need it, tho'.

Comment (workingjubilee) [reply pgx/src/array.rs:165]: Okaaaay it was just fun to write.

**Pull request merged**
